server/text_generation_server/models/flash_causal_lm.py

```diff
@@ -390,9 +407,20 @@ def generate_token(
             past_key_values,
         )
 
+        # Initialize past padding tensor
+        if self.past_pad is None:
+            self.past_pad = present.new_zeros(present.shape[0], 1, *present.shape[2:])
+
         # Initialize past_key_values in prefill
         if batch.past_key_values is None:
-            batch.past_key_values = [None] * len(batch)
+            batch.past_pad = self.past_pad
+            if len(batch) == 1:
+                # Preallocate tensor for bs = 1 case
+                batch.past_key_values = torch.nn.functional.pad(
+                    present, (0, 0, 0, 0, 0, 0, 0, batch.requests[0].stopping_parameters.max_new_tokens)
+                )
```

Comment:
We can avoid an additional allocation/copy here by tweaking the `forward` method of the various `flash_x_modeling` impls to optionally allocate a larger initial tensor.